### PR TITLE
Add list of connector plugins to Kafka Connect status

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- Enhancement / new feature

### Description

One of the things user needs to know to manage connectors using the Connector operator is which connectors are available in the Connect deployment. This PR adds list of available plugins to Kafka Connect and Kafka Connect S2I resources.

The status with this change looks something like this:

```
status:
  buildConfigName: my-connect-cluster-connect
  conditions:
  - lastTransitionTime: "2020-01-18T17:43:26.736Z"
    status: "True"
    type: Ready
  connectorPlugins:
  - class: cz.scholz.kafka.connect.echosink.EchoSinkConnector
    type: sink
    version: 1.0.0
  - class: org.apache.kafka.connect.file.FileStreamSinkConnector
    type: sink
    version: 2.4.0
  - class: org.apache.kafka.connect.file.FileStreamSourceConnector
    type: source
    version: 2.4.0
  - class: org.apache.kafka.connect.mirror.MirrorCheckpointConnector
    type: source
    version: "1"
  - class: org.apache.kafka.connect.mirror.MirrorHeartbeatConnector
    type: source
    version: "1"
  - class: org.apache.kafka.connect.mirror.MirrorSourceConnector
    type: source
    version: "1"
  observedGeneration: 1
  url: http://my-connect-cluster-connect-api.myproject.svc:8083
```

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java [hunk 1]
```diff
@@ -212,4 +215,37 @@ public KafkaConnectApiImpl(Vertx vertx) {
...
+
+                            try {
+                                result.complete(Arrays.asList(mapper.readValue(buffer.getBytes(), ConnectorPlugin[].class)));
+                            } catch (IOException e)  {
+                                log.warn("Failed to parse list of connector plugins");
+                                result.fail(e);
+                            }
+                        });
+                    } else {
+                        result.fail("Unexpected status code " + response.statusCode()
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java]: There's `ConnectRestException` which you should use here.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java [hunk 2]
```diff
@@ -212,4 +215,37 @@ public KafkaConnectApiImpl(Vertx vertx) {
...
+                    });
+                    if (response.statusCode() == 200) {
+                        response.bodyHandler(buffer -> {
+                            ObjectMapper mapper = new ObjectMapper();
+
+                            try {
+                                result.complete(Arrays.asList(mapper.readValue(buffer.getBytes(), ConnectorPlugin[].class)));
+                            } catch (IOException e)  {
+                                log.warn("Failed to parse list of connector plugins");
+                                result.fail(e);
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApi.java]: I wonder if you should use `ConnectRestException` here. Until now `ConnectRestException` has been for lower level errors in handling the HTTP. Here the HTTP worked but returned something that wasn't valid JSON, or was not the expected JSON.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiTest.java [hunk 3]
```diff
@@ -96,7 +98,22 @@ public void after() {
...
+
+                ConnectorPlugin fileSink = connectorPlugins.stream().filter(connector -> "org.apache.kafka.connect.file.FileStreamSinkConnector".equals(connector.getConnectorClass())).findFirst().orElse(null);
+                assertNotNull(fileSink);
+                assertEquals(fileSink.getType(), "sink");
+                assertEquals(fileSink.getVersion(), "2.3.1");
+
+                ConnectorPlugin fileSource = connectorPlugins.stream().filter(connector -> "org.apache.kafka.connect.file.FileStreamSourceConnector".equals(connector.getConnectorClass())).findFirst().orElse(null);
+                assertNotNull(fileSource);
+                assertEquals(fileSource.getType(), "source");
+                assertEquals(fileSource.getVersion(), "2.3.1");
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectApiTest.java]: So we really want to assert on the exact version number? We'll end up having to do this every time we change the version of Kafka being used to run the test. I think it's enough to test non-emptiness.

**Comment by scholzj** [comment]
@tombentley Should be all recolved now.

**Pull request merged**